### ISSUE TITLE
ROL: ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2 test failed sporadically

### ISSUE THREAD

**Issue by sebrowne** [trilinos/Trilinos#11621]
## Question
@trilinos/rol 
I saw ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2 fail twice last night in the newly-reenabled Intel 2021.3.0/OpenMPI 4.0.5 PR build.  I wanted to ask if there was any possibility if this was a code error, or if we should suspect an issue with the MPI installation/system load/other non-code issue?  I recognize that this is very difficult to answer definitively, so a guess is fine.  Just trying to decide how to handle it if it starts to fail more frequently.

Here are the two specific failure instances:
[ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2](https://trilinos-cdash.sandia.gov/test/20923276)
[ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2](https://trilinos-cdash.sandia.gov/test/20948962)

**Comment by sebrowne** [comment]
I actually just saw it on #11614 as well, but this time the CUDA build (still OpenMPI 4.0.5 though): [ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2](https://trilinos-cdash.sandia.gov/test/21003042)

**Comment by dridzal** [comment]
@sebrowne :  If the failure is random, I'm not sure what to do.  @eric-c-cyr , do you think this could be a code issue?

**Comment by sebrowne** [comment]
I know this is a tall ask because it's not usually failing.  I only noticed it because it was the same test a couple of times.  Figured I would ask in case it was obviously related to any recent ROL work, but by no means do I think that's the only possibility here.  I'll see if I can reproduce it on my blade if I run the test in a loop (which is still not fun, but possibly may be debuggable, especially if I can build my own MPI and that solves the issue).

**Comment by sebrowne** [comment]
Okay, I have an update!  Using this script:
```
#!/bin/bash

for i in {1..100}
do
    echo -e "\n#############\nITERATION ${i}\n#############"
    ctest -V --timeout 10 -R ROL_example_PinT_parabolic-control_AugmentedSystem_test_MPI_2 || exit 1
done
```
and this genconfig build: `rhel7_sems-intel-2021.3-sems-openmpi-4.0.5_release-debug_shared_no-kokkos-arch_no-asan_no-complex_fpic_mpi_no-pt_no-rdc_no-uvm_deprecated-on_all`

I can replicate the failure!  It appears to me that the change that caused it was #11615.  If I remove --bind-to none from the test command, it doesn't fail in 100 iterations.  If it's there, it fails on average within 10 iterations.

The kicker is that I can't replicate it on my blade, but I CAN replicate it on a Skylake compute machine, as well as a Broadwell.  I asked SEMS and their build machine is a Cascade Lake (like my blade).

I then tried building my open copy of OpenMPI 4.0.5 on the Broadwell machine I was using to test, and it cleared the error (even while still using --bind-to none).


Summary: something is behaving naughtily when using this combination of things:
1. Using a pre-cascade lake CPU machine
2. Using SEMS OpenMPI (I suspect this is the same issue as #11530, but there we were using OpenMPI 1.10.7)
3. Using the --bind-to none flag (which we need for running our tests in parallel)

All that being said, I don't know what the immediate fix is.  I'm loathe to revert the core binding change, because that makes all of the tests take far longer, and frequently many of them time out, and overall will affect the autotester more than just this one failing.  Since we've only seen the failure in this one specific test, I suppose I'm in favor of disabling it for Intel and CUDA until Framework can diagnose this with SEMS (currently there is a story in our backlog third from the top to work this).

**Comment by cgcgcg** [comment]
@sebrowne When running nightly tests on my workstation, `--bind-to hwthread` seems to work well. Doesn't mean that it will work well for Trilinos, but maybe it's worth a shot to test it with the failing ROL driver and one of slow nightly builds?

**Comment by sebrowne** [comment]
I tried that binding, and unfortunately it still maps multiple ranks to the same core, and they appear to bind there.  This is probably less of an issue with low -j values to ctest, but when I tested it I can see the processes all landing and binding on the same cores still :(

Without really controlling the mapping (using the ctest resource management, for instance), I think we're pretty much reliant on letting the OS manage CPU load across different cores, which requires --bind-to none.

**Comment by sebrowne** [comment]
More on this: I was able to replicate the sporadic failure if I built OpenMPI with spack on my blade and then run it on a Broadwell compute machine.  That is, unless I use the "internal-hwloc" variant (and if I build it without spack, which also uses the internal hwloc, still passes).  So, I tentatively think it's something about how the spack hwloc package is being built.

**Comment by sebrowne** [comment]
Further info:
This spack spec fails: `spack install openmpi@4.0.5+gpfs %intel@2021.3.0 arch=x86_64`
This spack spec passes: `spack install openmpi@4.0.5+gpfs %intel@2021.3.0 ^hwloc~libxml2 arch=x86_64`

**Comment by GrahamBenHarper** [comment]
@dridzal hwloc is a C package which helps bind software to hardware, and it required by a number of parallel packages like openmpi and tbb. Its supposed to be a more general/abstract software that covers a wide range of architectures, and it does things like managing the hardware topology and binding parallel software to physical cores.

**Comment by jhux2** [comment]
Also affecting #11661.

**Comment by jhux2** [comment]
@trilinos/rol Would it be ok to disable this test temporarily?

**Comment by dridzal** [comment]
@jhux2 :  I think that would be ok ... can you do that just for these platforms, or do we have to disable it fully on the ROL side?

**Comment by jhux2** [comment]
> I think that would be ok ... can you do that just for these platforms, or do we have to disable it fully on the ROL side?

@sebrowne Is this something that can be done on a per-platform basis for the PR testing?

**Comment by sebrowne** [comment]
Yes we can.  It needs to be disabled for each of the OpenMPI 4.0.X platforms (so Intel and CUDA).  We're working with SEMS on getting the new toolchain deployed that will fix these long-term.  Note that it may affect other tests, this one just seems particularly sensitive to the issue.  You can even do it in the failing MR (see packages/framework/config_specs.ini).  I can take care of it later this evening on #11654 if you'd like.

**Comment by jhux2** [comment]
@sebrowne It's probably better if I don't touch that file ....  I'm perfectly fine waiting until tonight.

**Comment by sebrowne** [comment]
Here is the diff to include to disable that test: https://github.com/trilinos/Trilinos/compare/develop...sebrowne:Trilinos:develop

**Comment by sebrowne** [comment]
Given that this has been disabled (temporarily), and that we're fairly certain it's an issue with MPI, not with ROL, I'm going to go ahead and close this issue.
